File: src/main/java/com/freya02/botcommands/internal/application/slash/autocomplete/suppliers/ChoiceSupplierStringFuzzy.java
package com.freya02.botcommands.internal.application.slash.autocomplete.suppliers;

import com.freya02.botcommands.internal.application.slash.SlashCommandInfo;
import com.freya02.botcommands.internal.application.slash.autocomplete.AutocompletionHandlerInfo;
import com.freya02.botcommands.internal.application.slash.autocomplete.ChoiceSupplier;
import me.xdrop.fuzzywuzzy.FuzzySearch;
import me.xdrop.fuzzywuzzy.model.ExtractedResult;
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.Command;
import net.dv8tion.jda.api.interactions.commands.OptionMapping;

import java.util.Collection;
import java.util.List;

public class ChoiceSupplierStringFuzzy implements ChoiceSupplier {
	private final AutocompletionHandlerInfo handlerInfo;

	public ChoiceSupplierStringFuzzy(AutocompletionHandlerInfo handlerInfo) {
		this.handlerInfo = handlerInfo;
	}

	@Override
	public List<Command.Choice> apply(SlashCommandInfo slashCommand, CommandAutoCompleteInteractionEvent event, Collection<?> collection) throws Exception {
		final List<String> list = collection
				.stream()
				.map(Object::toString)
				.sorted()
				.toList();

		final OptionMapping optionMapping = event.getFocusedOption();
		//First sort the results by similarities but by taking into account an incomplete input
		final List<ExtractedResult> bigLengthDiffResults = FuzzySearch.extractTop(optionMapping.getAsString(),
				list,
				FuzzySearch::partialRatio,
				handlerInfo.getMaxChoices());

		//Then sort the results by similarities but don't take length into account
		final List<ExtractedResult> similarities = FuzzySearch.extractTop(optionMapping.getAsString(),
				bigLengthDiffResults.stream().map(ExtractedResult::getString).toList(),
				FuzzySearch::ratio,
				handlerInfo.getMaxChoices());

		return similarities.stream()
				.limit(handlerInfo.getMaxChoices())
				.map(c -> AutocompletionHandlerInfo.getChoice(optionMapping, c.getString()))
				.toList();
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/internal/commands/application/slash/autocomplete/suppliers/ChoiceSupplierStringFuzzy.kt
package io.github.freya022.botcommands.internal.commands.application.slash.autocomplete.suppliers

import io.github.freya022.botcommands.api.commands.application.slash.autocomplete.AutocompleteAlgorithms
import io.github.freya022.botcommands.internal.commands.application.slash.autocomplete.AutocompleteHandler.Companion.asChoice
import io.github.freya022.botcommands.internal.utils.throwArgument
import net.dv8tion.jda.api.events.interaction.command.CommandAutoCompleteInteractionEvent
import net.dv8tion.jda.api.interactions.commands.Command

internal class ChoiceSupplierStringFuzzy(private val numChoices: Int) : ChoiceSupplier {
    @Throws(Exception::class)
    override fun apply(event: CommandAutoCompleteInteractionEvent, collection: Collection<Any>): List<Command.Choice> {
        val autoCompleteQuery = event.focusedOption
        return AutocompleteAlgorithms.fuzzyMatching(collection, { it.toString() }, event.focusedOption.value)
            .take(numChoices)
            .map { it.string.asChoice(autoCompleteQuery.type) ?: throwArgument("Malformed input for option type ${autoCompleteQuery.type}: '${it.string}'") }
    }
}